vermouth/processors/annotate_mut_mod.py

```diff
@@ -0,0 +1,202 @@
+        A graph with one node per residue.
+    res_idx: collections.abc.Hashable
+        A node index in residue_graph.
+
+    Returns
+    -------
+    bool
+        Whether resspec describes the node res_idx in residue_graph.
+    """
+    res_node = residue_graph.nodes[res_idx]
+    residue = {key: res_node.get(key)
+               for key in 'chain resid resname insertion_code'.split()}
+    out = True
+    if resspec.get('resname', '')[-3:] == 'ter':
+        # Find all residues with degree 1: the ones with the lowest resid will
+        # be cter, the ones with the highest resid nter.
+        termini = [idx for idx in residue_graph if residue_graph.degree[idx] == 1]
+        get_resid = lambda idx: residue_graph.nodes[idx].get('resid')
+        # FIXME: Once residue_graph is a digraph we can do something much much
+        #        more clever, addressing arbitrarily branched polymers and
+        #        termini
+        if resspec['resname'] == 'nter':
+            return res_idx in maxes(termini, key=lambda x: -get_resid(x))
+        elif resspec['resname'] == 'cter':
+            return res_idx in maxes(termini, key=get_resid)
+        else:
+            raise KeyError("Don't know any terminus with name '{}'".format(resspec['resname']))
+    return out and _subdict(resspec, residue)
+
+
+def _format_resname(res):
+    """
+    Provisional function that performs the opposite of parse_residue_spec.
+    Poorly tested, use at own risk.
+    """
+    chain = res.get('chain', '')
+    out = ''
+    if chain:
+        out += chain + '-'
+    resname = res.get('resname')
+    out += resname
+    if resname and resname[-1].isdigit():
+        out += '#'
+    out += str(res.get('resid', ''))
+    out += res.get('insertion_code', '')
+    return out
+
+
+def annotate_modifications(molecule, modifications, mutations):
+    """
+    Annotate nodes in molecule with the desired modifications and mutations
+
+    Parameters
+    ----------
+    molecule: networkx.Graph
+    modifications: list[tuple[dict, str]]
+        The modifications to apply. The first element is a dictionary contain
+        the attributes a residue has to fulfill. It can contain the elements
+        'chain', 'resname' and 'resid'. The second element is the modification
+        that should be applied.
+    mutations: list[tuple[dict, str]]
+        The mutations to apply. The first element is a dictionary contain
+        the attributes a residue has to fulfill. It can contain the elements
+        'chain', 'resname' and 'resid'. The second element is the mutation that
+        should be applied.
+
+    Raises
+    ------
+    NameError
+        When a modification is not recognized.
+    """
+    if not modifications and not mutations:
+        return
+
+    residue_graph = make_residue_graph(molecule)
+    for res_idx in residue_graph:
+        for mutmod, key, library in [(modifications, 'modification', molecule.force_field.modifications),
```

Comment:
It can't be moved to the top of the file since it depends on the function arguments. I did give it a name and moved it higher in the function. In addition, added a small comment.